File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# First Solution = recursive DFS, O(n)
# class Solution:
#     def maxDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0
        
#         return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1;
    
# Second Solution = iterative BFS, O(n)
# deque = "double ended queue" - doubly linked list that gives O(1) pop and push for head and tail
# from collections import deque

# class Solution:
#     def maxDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0
        
#         level = 0
#         queue = deque([root])
#         while queue:
#             for i in range(len(queue)):
#                 node = queue.popleft()
#                 if node.left:
#                     queue.append(node.left)
#                 if node.right:
#                     queue.append(node.right)
#             level += 1
            
#         return level
        
# Third Solution = iterative DFS, O(n) - preorder easiest with iterative
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = [[root, 1]]
        level = 0
        
        while stack:
            node, depth = stack.pop()
            if node:
                level = max(level, depth)
                stack.append([node.right, depth + 1])
                stack.append([node.left, depth + 1])

        return level

File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// Recursive DFS O(n), 
// var maxDepth = function(root) {
//     if (!root) return null;
    
//     return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
// };

// iterative DFS
var maxDepth = function(root) {
    if (!root) return 0
    
    let stack = [[root, 1]];
    let level = 0;

    while (stack.length) {
        let node = stack.pop();
        
        if (node !== null) {
            level = Math.max(level, node[1]);
            if (node[0].left !== null) stack.push([node[0].left, node[1] + 1]);
            if (node[0].right !== null) stack.push([node[0].right, node[1] + 1]);
        }
    }
    
    return level
};